wallet: detecting duplicate wallet by comparing the db filename.

ken2812221 opened the pull request:
Fix #14538

Fix crash attempting to load the same wallet with different path strings that resolve to the same absolute path. The primary check which prevents loading the same wallet twice is:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/db.cpp#L44

But this check is skipped if both wallet paths resolve to the same absolute path, due to caching here:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/db.cpp#L467

Meanwhile a secondary check for duplicate wallets is not reliable because it based on a literal comparison, instead of comparison using absolute paths:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/wallet.cpp#L3853

This PR fixes the latter check to compare the absolute path of a new wallet being loaded to absolute paths of wallets already loaded, so there should no longer be any way to load the same wallet more than once.

ken2812221 committed:
wallet: Add trailing wallet.dat when detecting duplicate wallet if it's a directory.

ryanofsky commented:
Would prefer to check this at database level, rather than in higher level wallet code. I think it's better if high level code just passes along wallet paths and isn't concerned with how data is stored in them.

I posted a tweaked version of this PR at 40a6475e59950838be532b0a2c58f4359eae5625 which moves the check. @ken2812221, could you incorporate 40a6475e59950838be532b0a2c58f4359eae5625 into this PR if you think it makes sense?

ken2812221 commented:
@ryanofsky The element in mapDb can be loaded lazily. 
For example: If it is a new wallet or the other wallet in same directory just unloaded, the element in mapDb dies not exist.
So the only vector that store the wallets we load is `vpwallets`.

ryanofsky committed:
Refactor: Move m_db pointers into BerkeleyDatabase

This is a refactoring change that doesn't affect behavior. The motivation
behind the change is give BerkeleyEnvironment objects access to
BerkeleyDatabase objects so it will be possible to simplify the duplicate
wallet check and more reliably avoid opening the same databases twice.

ryanofsky commented:
> @ryanofsky The element in mapDb is not always exist.

Good catch. I think a small refactoring could fix this: 9c945d011abde24f8649d877e694d392e1a5727e. And then the fix would change slightly to aef19be5851ceb3b324fd80756060a00aad8e61c.

I pushed a branch with these two commits to: https://github.com/ryanofsky/bitcoin/commits/pr/bloaded

I think this is a better than checking for wallet.dat files outside the database layer, but let me know what you think, and feel free to use the code or commits in this PR if you think this approach makes sense.

ryanofsky reviewed (approved):
utACK aef19be5851ceb3b324fd80756060a00aad8e61c

ryanofsky reviewed (approved):
utACK b0375538375eb0c419b308d1015723ab64d53294. Only change since last review is rebase.

@promag, this might be an easy PR for to you review. First commit is just refactoring that doesn't change behavior. Second commit is the actual fix.

@ken2812221, it might help for this to have a more complete PR description.

---

Suggested PR description:

Fix crash attempting to load the same wallet with different path strings that resolve to the same absolute path. The primary check which prevents loading the same wallet twice is:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/db.cpp#L44

But this check is skipped if both wallet paths resolve to the same absolute path, due to caching here:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/db.cpp#L467

Meanwhile a secondary check for duplicate wallets is not reliable because it based on a literal comparison, instead of comparison using absolute paths:

https://github.com/bitcoin/bitcoin/blob/6b8d0a2164b30eab76e7bccb1ffb056a10fba406/src/wallet/wallet.cpp#L3853

This PR fixes the latter check to compare the absolute path of a new wallet being loaded to absolute paths of wallets already loaded, so there should no longer be any way to load the same wallet more than once.

promag commented on src/wallet/db.cpp:
```diff
@@ -463,7 +463,8 @@ BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bo
             throw std::runtime_error("BerkeleyBatch: Failed to open database environment.");
 
-        pdb = env->mapDb[strFilename];
+        BerkeleyDatabase& database = env->m_databases.at(strFilename).get();
```
Some comments here:
 - there's a slight change when replacing `map::operator[]` with `map::at`, maybe worth a comemnt or an assertion?
 - also, isn't this equivalent to the above `database` argument?
 - is this is really necessary then you could avoid shadowing `database`.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14552#discussion_r230800120

Oh, I didn't even realize there was a `database` argument being shadowed above. Should just delete this line and use the existing `database`, which points to the same thing.

promag commented on src/wallet/db.h:
```diff
@@ -115,13 +117,22 @@ class BerkeleyDatabase
         }
     }
 
+    ~BerkeleyDatabase() {
```
nit, could add `final` to class since this is not virtual.

ken2812221 replied:
`final` can only be added to virtual method.

promag replied:
@ken2812221 I mean `class BerkeleyDatabase final`.

ryanofsky replied:
Change would be unrelated to this PR, and IMO, it's better only to use final when you have virtual methods and actual optimizations this would allow or bugs it would prevent. There's a core guideline that touches on this: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-final

promag replied:
Nice thanks.

ken2812221 commented:
@ryanofsky Thanks for your review. I would take your suggestion.

promag commented on src/wallet/db.cpp:
```diff
@@ -56,9 +56,8 @@ bool WalletDatabaseFileId::operator==(const WalletDatabaseFileId& rhs) const
 }
 
-BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& database_filename)
+void SplitWalletPath(const fs::path& wallet_path, fs::path& env_directory, std::string& database_filename)
```
Could be `static`?

promag commented on src/wallet/db.cpp:
```diff
@@ -563,12 +581,11 @@ void BerkeleyEnvironment::CloseDb(const std::string& strFile)
     {
         LOCK(cs_db);
-        if (mapDb[strFile] != nullptr) {
+        BerkeleyDatabase& database = m_databases.at(strFile).get();
```
Same as above, there's a slight change when replacing map::operator[] with map::at, maybe worth a comment or an assertion?

promag commented:
utACK 15c93f0.

laanwj commented on src/wallet/db.cpp:
```diff
@@ -71,6 +70,24 @@ BerkeleyEnvironment* GetWalletEnv(const fs::path& wallet_path, std::string& data
+    LOCK(cs_db);
+    auto env = g_dbenvs.find(env_directory.string());
+    if (env == g_dbenvs.end()) return false;
+    auto db = env->second.m_databases.find(database_filename);
```
I think this is leaking the implementation detail of `m_databases`, wouldn't checking whether a database exists ideally be a responsibility of the database environment. e.g. add a `IsDatabaseLoaded` on `DBenv`?

promag replied:
Note that `BerkeleyEnvironment::mapDb` is currently used outside. I consider this a bugfix so any improvement/refactor could come next.

ryanofsky replied:
Resolved by adding `IsDatabaseLoaded`

ken2812221 committed:
wallet: Create IsDatabaseLoaded function

ryanofsky commented:
utACK 591203149f1700f594f781862e88cbbfe83d8d37. Just minor tweaks since last review: replacing map::at exception with assert, adding IsDatabaseLoaded.

@promag and @MarcoFalke, I've noticed that in your recent reviews you've been advocating replacing exceptions with asserts or return codes. Maybe one of you can write up some guidelines about when to use exceptions in the developer guide. Personally, I tend to like exceptions and think that whatever drawbacks they have will be fixed by new c++ features like `throws`/`std::error` in 
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf (see §4.1.1 "Elevator Pitch")

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/14552/commits/591203149f1700f594f781862e88cbbfe83d8d37, looks good

ryanofsky commented:
Can this be merged? It fixes an ugly bug and has some refactoring changes I want to build on. If it can't be merged, maybe it could be added to https://github.com/bitcoin/bitcoin/projects/8

MarcoFalke commented:
Is this for backport?

promag commented on src/wallet/db.h:
```diff
@@ -43,7 +45,7 @@ class BerkeleyEnvironment
     std::unique_ptr<DbEnv> dbenv;
     std::map<std::string, int> mapFileUseCount;
-    std::map<std::string, Db*> mapDb;
+    std::map<std::string, std::reference_wrapper<BerkeleyDatabase>> m_databases;
```
Shouldn't this be guarded by `cs_db`?
